src/amltk/optimization/optimizers/optuna.py

```diff
@@ -292,7 +292,36 @@ def ask(
         if n is not None:
             return (self.ask(n=None) for _ in range(n))
 
-        optuna_trial: optuna.Trial = self.study.ask(self.space)
+        if any("__choice__" in k for k in self.space):
+            optuna_trial: optuna.Trial = self.study.ask()
+            # do all __choice__ suggestions with suggest_categorical
+            workspace = self.space.copy()
+            delete_other_options = []
+            for name, distribution in workspace.items():
+                if "__choice__" in name:
+                    possible_choices = distribution.choices
+                    choice_made = optuna_trial.suggest_categorical(name, choices=possible_choices)
+                    for c in possible_choices:
+                        if c != choice_made:
+                            delete_other_options.append(f"{name}:{c}:")
+            # filter all parameters given the made choices
+            filtered_workspace = {k: v for k, v in workspace.items() if (
+                ("__choice__" not in k) and
+                (not any(c in k for c in delete_other_options))
```

Comment:
This line might need a second look, I had a bug a while back with something similar where essentially the `c` was present in some other `k` in another part of the pipeline that shouldn't have been effected by the given choice.

I'm on my phone so I cant write a concrete example but see the overall review comment for more